File: 22.4/ex1.rb
# encoding: UTF-8
STDIN.set_encoding(Encoding::UTF_8)
STDOUT.set_encoding(Encoding::UTF_8)

class CanBo
    attr_reader :name, :birth_year, :gender, :address
  
    def initialize(name, birth_year, gender, address)
      @name = name
      @birth_year = birth_year
      @gender = gender
      @address = address
    end
  
    def display_info
      raise NotImplementedError, 'You must implement the display_info method'
    end
  end
  
  class NhanVien < CanBo
    attr_reader :job
  
    def initialize(name, birth_year, gender, address, job)
      super(name, birth_year, gender, address)
      @job = job
    end
  
    def display_info
      "NhanVien{name=#{name}, birthYear=#{birth_year}, gender=#{gender}, address=#{address}, job=#{job}}"
    end
  end
  
  class CongNhan < CanBo
    attr_reader :level
  
    def initialize(name, birth_year, gender, address, level)
      super(name, birth_year, gender, address)
      @level = level
    end
  
    def display_info
      "CongNhan{name=#{name}, birthYear=#{birth_year}, gender=#{gender}, address=#{address}, level=#{level}}"
    end
  end
  
  class KySu < CanBo
    attr_reader :major
  
    def initialize(name, birth_year, gender, address, major)
      super(name, birth_year, gender, address)
      @major = major
    end
  
    def display_info
      "KySu{name=#{name}, birthYear=#{birth_year}, gender=#{gender}, address=#{address}, major=#{major}}"
    end
  end
  
  class QLCB
    def initialize
      @can_bo_list = []
    end
  
    def add_can_bo(can_bo)
      @can_bo_list << can_bo
    end
  
    def search_by_name(name)
      @can_bo_list.select { |can_bo| can_bo.name == name }
    end
  
    def display_all
      @can_bo_list.each { |can_bo| puts can_bo.display_info }
    end
  end
  
  qlcb = QLCB.new
  exit = false
  
  while !exit
    puts '1. Nhap thong tin moi cho can bo'
    puts '2. Tim kiem theo ho ten'
    puts '3. Hien thi thong tin danh sach can bo'
    puts '4. Thoat chuong trinh'
    print 'Nhap lua chon cua ban: '
    choice = gets.chomp.to_i
  
    case choice
    when 1
      print 'Nhap loai can bo (NhanVien, CongNhan, KySu): '
      type = gets.chomp
  
      print 'Nhap ten: '
      name = gets.chomp
  
      print 'Nhap nam sinh: '
      birth_year = gets.chomp.to_i
  
      print 'Nhap gioi tinh: '
      gender = gets.chomp
  
      print 'Nhap dia chi: '
      address = gets.chomp
  
      case type.downcase
      when 'nhanvien'
        print 'Nhap nghe nghiep: '
        job = gets.chomp
        qlcb.add_can_bo(NhanVien.new(name, birth_year, gender, address, job))
      when 'congnhan'
        print 'Nhap cap bac: '
        level = gets.chomp.to_i
        qlcb.add_can_bo(CongNhan.new(name, birth_year, gender, address, level))
      when 'kysu'
        print 'Nhap nganh dao tao: '
        major = gets.chomp
        qlcb.add_can_bo(KySu.new(name,  birth_year, gender, address, major))
    else
      puts 'Sai loai can bo!'
    end
    when 2
        print 'Nhap ten de tim kiem: '
            search_name = gets.chomp
            search_result = qlcb.search_by_name(search_name)
                if search_result.size > 0
                    search_result.each { |can_bo| puts can_bo.display_info }
            else
        puts 'Khong co canbo nao co ten nhu vay.'
            end
    when 3
        qlcb.display_all
    when 4
        exit = true
            else
        puts 'Lua chon khong hop le, vui long thu lai.'
    end
end    
  

File: 22.4/ex1.java
import java.util.ArrayList;
import java.util.Scanner;

abstract class CanBo {
    private String name;
    private int birthYear;
    private String gender;
    private String address;

    public CanBo(String name, int birthYear, String gender, String address) {
        this.name = name;
        this.birthYear = birthYear;
        this.gender = gender;
        this.address = address;
    }

    public String getName() {
        return name;
    }

    public int getBirthYear() {
        return birthYear;
    }

    public String getGender() {
        return gender;
    }

    public String getAddress() {
        return address;
    }

    public abstract String displayInfo();
}

class NhanVien extends CanBo {
    private String job;

    public NhanVien(String name, int birthYear, String gender, String address, String job) {
        super(name, birthYear, gender, address);
        this.job = job;
    }

    public String getJob() {
        return job;
    }

    @Override
    public String displayInfo() {
        return "NhanVien{" +
                "name=" + getName() +
                ", birthYear=" + getBirthYear() +
                ", gender=" + getGender() +
                ", address=" + getAddress() +
                ", job=" + job +
                '}';
    }
}

class CongNhan extends CanBo {
    private int level;

    public CongNhan(String name, int birthYear, String gender, String address, int level) {
        super(name, birthYear, gender, address);
        this.level = level;
    }

    public int getLevel() {
        return level;
    }

    @Override
    public String displayInfo() {
        return "CongNhan{" +
                "name=" + getName() +
                ", birthYear=" + getBirthYear() +
                ", gender=" + getGender() +
                ", address=" + getAddress() +
                ", level=" + level +
                '}';
    }
}

class KySu extends CanBo {
    private String major;

    public KySu(String name, int birthYear, String gender, String address, String major) {
        super(name, birthYear, gender, address);
        this.major = major;
    }

    public String getMajor() {
        return major;
    }

    @Override
    public String displayInfo() {
        return "KySu{" +
                "name=" + getName() +
                ", birthYear=" + getBirthYear() +
                ", gender=" + getGender() +
                ", address=" + getAddress() +
                ", major=" + major +
                '}';
    }
}

class QLCB {
    private ArrayList<CanBo> canBoList = new ArrayList<>();

    public void addCanBo(CanBo canBo) {
        canBoList.add(canBo);
    }

    public ArrayList<CanBo> searchByName(String name) {
        ArrayList<CanBo> result = new ArrayList<>();
        for (CanBo canBo : canBoList) {
            if (canBo.getName().equals(name)) {
                result.add(canBo);
            }
        }
        return result;
    }

    public void displayAll() {
        for (CanBo canBo : canBoList) {
            System.out.println(canBo.displayInfo());
        }
    }
}

public class ex1 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        QLCB qlcb = new QLCB();
        boolean exit = false;

        while (!exit) {
            System.out.println("1. Nhap thong tin moi cho can bo");
            System.out.println("2. Tim kiem theo ho ten");
            System.out.println("3. Hien thi thong tin danh sach can bo");
            System.out.println("4. Thoat chuong trinh");
            System.out.print("Nhap lua chon cua ban: ");
            int choice = scanner.nextInt();
            scanner.nextLine();
    
            switch (choice) {
                case 1:
                    System.out.print("Nhap loai can bo (NhanVien, CongNhan, KySu): ");
                    String type = scanner.nextLine();
    
                    System.out.print("Nhap ten: ");
                    String name = scanner.nextLine();
    
                    System.out.print("Nhap nam sinh: ");
                    int birthYear = scanner.nextInt();
                    scanner.nextLine();
    
                    System.out.print("Nhap gioi tinh: ");
                    String gender = scanner.nextLine();
    
                    System.out.print("Nhap dia chi: ");
                    String address = scanner.nextLine();
    
                    if (type.equalsIgnoreCase("NhanVien")) {
                        System.out.print("Nhap nghe nghiep: ");
                        String job = scanner.nextLine();
                        qlcb.addCanBo(new NhanVien(name, birthYear, gender, address, job));
                    } else if (type.equalsIgnoreCase("CongNhan")) {
                        System.out.print("Nhap cap bac: ");
                        int level = scanner.nextInt();
                        scanner.nextLine();
                        qlcb.addCanBo(new CongNhan(name, birthYear, gender, address, level));
                    } else if (type.equalsIgnoreCase("KySu")) {
                        System.out.print("Nhap nganh dao tao: ");
                        String major = scanner.nextLine();
                        qlcb.addCanBo(new KySu(name, birthYear, gender, address, major));
                    } else {
                        System.out.println("Sai loai can bo!");
                    }
                    break;
                case 2:
                    System.out.print("Nhap ten de tim kiem: ");
                    String searchName = scanner.nextLine();
                    ArrayList<CanBo> searchResult = qlcb.searchByName(searchName);
                    if (searchResult.size() > 0) {
                        for (CanBo canBo : searchResult) {
                            System.out.println(canBo.displayInfo());
                        }
                    } else {
                        System.out.println("Khong co canbo nao co ten nhu vay.");
                    }
                    break;
                case 3:
                    qlcb.displayAll();
                    break;
                case 4:
                    exit = true;
                    break;
                default:
                    System.out.println("Lua chon khong hop le, vui long thu lai.");
            }
        }
        scanner.close();
    }
}
    
